### PR TITLE
[wasm] pinvoke improvements part 2

### PR DESCRIPTION
**Pull request by kg**
Expands WBT test coverage and makes the following improvements:
* InlineArray is supported in wasm pinvokes, and in general works correctly in mono marshaling now (it was broken on all targets)
* fixed arrays are supported in wasm pinvokes
* struct arguments and return values are supported in more wasm pinvoke scenarios (not all, though)
* wasm struct scalarization is more intelligent about noticing padding and size mismatches in structs (without this, some inlinearrays would be scalarized incorrectly)
* Prevent infinite recursion in build task's TypeToChar

### CODE REVIEW

**Comment by kg** [comment]
cc @vargaz there is some debugging code in here now, the problem i'm digging into is that System.Runtime.Tests fails in AOT mode.
Part of the the problem appears to be that Assert.Equals drops into the interpreter, and then at some point we call Task<T>.GetAwaiter from the interpreter, and the jitcall to do that appears to use the wrong signature - it expects (ii) -> i, and it seems like the target is (iiiii) or (iii), so there's probably a disagreement somewhere on whether TaskAwaiter gets scalarized. TaskAwaiter has one field of type Task so it probably should be scalarized, but we might not be allowing that for reftypes in all the scalar vtype code we wrote. I'm not sure whether scalarizing a GC reference is safe since it wouldn't be visible to the collector anymore.

File: src/mono/mono/mini/mini-wasm.c [hunk 1]
```diff
@@ -75,17 +75,23 @@ get_storage (MonoType *type, MonoType **etype, gboolean is_return)
 	case MONO_TYPE_R8:
 		return ArgOnStack;
 
-	case MONO_TYPE_GENERICINST:
+	case MONO_TYPE_GENERICINST: {
 		if (!mono_type_generic_inst_is_valuetype (type))
 			return ArgOnStack;
 
+		if (mini_wasm_is_scalar_vtype (type, etype))
```
Comment (vargaz) [inline src/mono/mono/mini/mini-wasm.c:82]: This should be below the mini_is_gsharedvt_variable_type () check.

File: src/mono/mono/mini/mini-wasm.c [hunk 2]
```diff
@@ -769,7 +775,12 @@ mini_wasm_is_scalar_vtype (MonoType *type, MonoType **etype)
 		if (nfields > 1)
 			return FALSE;
 		MonoType *t = mini_get_underlying_type (field->type);
-		if (MONO_TYPE_ISSTRUCT (t)) {
+		int field_size = mono_class_value_size (mono_class_from_mono_type_internal (t), NULL);
+		// inlinearray and fixed both work by having a single field that is bigger than its element type.
```
Comment (vargaz) [inline src/mono/mono/mini/mini-wasm.c:779]: This should be
 ```gint32 align;
int field_size = mono_type_size (t, &align);
```

File: src/tasks/WasmAppBuilder/SignatureMapper.cs [hunk 3]
```diff
@@ -51,19 +58,34 @@ internal static class SignatureMapper
...
                 c = 'I';
-            else if (t.IsEnum)
-                c = TypeToChar(t.GetEnumUnderlyingType(), log);
-            else if (t.IsPointer)
+            else if (t.IsEnum) {
+                Type underlyingType = t.GetEnumUnderlyingType();
+                if (underlyingType != t)
+                    c = TypeToChar(underlyingType, log, out _, ++depth);
+                else {
+                    log.Warning("WASM0064", $"Unsupported parameter type '{t.Name}'");
```
Comment (maraf) [inline src/tasks/WasmAppBuilder/SignatureMapper.cs:66]: When can this happen?

Comment (kg) [reply src/tasks/WasmAppBuilder/SignatureMapper.cs:66]: I don't know of any realistic scenario where unbounded recursion can occur in TypeToChar, but it broke the wasm perf measurements so I am doing everything I can to fix it. I can't repro it locally.

File: src/mono/mono/mini/interp/interp.c [hunk 4]
```diff
@@ -1473,11 +1477,36 @@ get_build_args_from_sig_info (MonoMemoryManager *mem_manager, MonoMethodSignatur
...
 		case MONO_TYPE_GENERICINST:
-			info->ret_pinvoke_type = PINVOKE_ARG_INT;
+#ifdef HOST_WASM
+			// This ISSTRUCT check is important, because the type could be an enum
+			if (MONO_TYPE_ISSTRUCT (info->ret_mono_type)) {
+				// The return type was already filtered previously, so if we get here
+				//  we're returning a struct byref instead of as a scalar
+				info->ret_pinvoke_type = PINVOKE_ARG_WASM_VALUETYPE_RESULT;
+				info->ilen++;
+			} else {
```
Comment (vargaz) [inline src/mono/mono/mini/interp/interp.c:1503]: Would better to move the } else { out of the ifdefs somehow, it could confuse indentation in editors.

**Comment by vargaz** [comment]
The runtime changes look ok to me.

**Review by lewing — APPROVED** [review]

**Pull request merged**